### ISSUE TITLE
[Question] Use value of `SELECT_RAW` as an insert node's default text

### ISSUE THREAD

**Issue by ejmastnak** [L3MON4D3/LuaSnip#511]
First, my appreciation for your work on this plugin.

I'm trying to switch from UltiSnips to pure Lua snippets and got stuck on this issue; I thought I would ask here after not finding a solution in `:help luasnip` or the Wiki.

My goal: use the visual selection, i.e. the contents of `SELECT_RAW`, in the same role as an insert node's default text. Here is a naive implementation that won't work, but hopefully gets the point across:

```lua
-- Surround selected text in parentheses
s({trig="("},
  {
    t("("),
    i(1, SELECT_RAW), -- This of course doesn't work, it's just to (hopefully) explain my goal
    t(")"),
  }),
```

In case it helps explain what I'm after, the equivalent (correctly working) UltiSnips snippet is:

```snippets
snippet ( "Paired parentheses"
(${1:${VISUAL:}})
endsnippet
```

Here is my LuaSnip attempt based on this example [function](https://github.com/L3MON4D3/LuaSnip/blob/master/Examples/snippets.lua#L179) and [snippet](https://github.com/L3MON4D3/LuaSnip/blob/master/Examples/snippets.lua#L290) from the docs: I use a dynamic node to interpolate the value of a function returning `parent.snippet.env.SELECT_RAW` into the initial value of an insert node snippet.

```lua
local get_visual_selection = function(args, parent)
  return sn(nil, i(1, parent.snippet.env.SELECT_RAW))
end

return
{
  s({trig="("},
    {
      t("("),
      d(1, get_visual_selection),
      t(")"),
    }),
}
```

This snippet *sort of* works---it correctly surrounds a visual selection in parentheses, but throws a `E5108 attempt to get length of a nil value` error when the trigger `(` is pressed with no active visual selection. (The full error stack and my LuaSnip config are below).
Here is a demo GIF:

![demo](https://user-images.githubusercontent.com/50270681/183388679-e3fe4454-05e3-4705-bfe2-62b44718e561.gif)

The `attempt to get length of a nil value` error makes me think `snippet.env.SELECT_RAW` is `nil`, so I tried adding a `nil` check like this...
```lua
local get_visual_selection = function(args, parent)
  if (parent.snippet.env.SELECT_RAW == nil) then
    return sn(nil, i(1, '')) -- return an empty string
  else
    return sn(nil, i(1, parent.snippet.env.SELECT_RAW))
  end
end
```
...but this still results in the `E5108` error. Which brings me to:

### Question

Could you suggest a clean way to use the value of `SELECT_RAW` as an insert node's default text such that snippet expands even when there is no active visual selection?

Apologies for such a verbose issue, I wanted to do my best to explain my goal clearly and show what I'd tried so far.

### Relevant config

```lua
ls.config.set_config({
	store_selection_keys = "<Tab>",
})
require("luasnip.loaders.from_lua").load({paths = "~/.config/nvim/LuaSnip"})
```

```lua
-- ~/.config/nvim/LuaSnip/all.lua
local get_visual_selection = function(args, parent)
  return sn(nil, i(1, parent.snippet.env.SELECT_RAW))
end

return
{
  s({trig="("},
    {
      t("("),
      d(1, get_visual_selection),
      t(")"),
    }),
}
```

### Error

```
E5108: Error executing lua ...nvim/plugged/LuaSnip/lua/luasnip/util/util.lua:233: attempt to get length of a nil value
stack traceback:
        ...nvim/plugged/LuaSnip/lua/luasnip/util/util.lua:233: in function 'put'
        ...nvim/plugged/LuaSnip/lua/luasnip/nodes/node.lua:63: in function 'put_initial'
        ...nvim/plugged/LuaSnip/lua/luasnip/nodes/snippet.lua:668: in function 'put_initial'
        ...nvim/plugged/LuaSnip/lua/luasnip/nodes/dynamicNode.lua:168: in function 'update'
        ...nvim/plugged/LuaSnip/lua/luasnip/nodes/snippet.lua:800: in function 'update'
        ...nvim/plugged/LuaSnip/lua/luasnip/nodes/snippet.lua:484: in function 'trigger_expand'
        ...nvim/plugged/LuaSnip/lua/luasnip/init.lua:220: in function 'snip_expand'
        ...nvim/plugged/LuaSnip/lua/luasnip/init.lua:274: in function 'expand'
        [string ":lua"]:1: in main chunk
```

**Comment by leiserfg** [comment]
You should be able to work around it by checking `#parent.snippet.env.SELECT_RAW > 0`  instead. But yes there is a bug either in the dynamic node or in the input node.

**Comment by leiserfg** [comment]
It's the input node, it requires (inexplicitly) that the default value is not an empty table, I'll rise a PR to fix it.

**Comment by ejmastnak** [comment]
Thanks for the super fast response! I can confirm that checking `#parent.snippet.env.SELECT_RAW > 0` solves my original issue.

> It's the input node, it requires (inexplicitly) that the default value is not an empty table, I'll rise a PR to fix it.

Great! In that case I'll leave this open until #512 is merged, but I would consider my problem solved at this point.

**Comment by L3MON4D3** [comment]
Okayy, just merged #512, I'll just go ahead and close :P

**Comment by ejmastnak** [comment]
Sounds good, thanks both @L3MON4D3 and @leiserfg for your help and fast responses!
